File: ios-paddle-up/PaddleUp/Models/Mechanics.swift
//
//  Mechanics.swift
//  PaddleUp
//
//  Mechanic taxonomy, per-shot rubrics (configurable weights) and benchmark
//  ranges. Benchmarks live here, apart from screen logic, so they can later be
//  replaced with values derived from coach review or elite footage.
//

import Foundation

nonisolated enum MechanicID: String, Codable, CaseIterable, Sendable, Identifiable {
    // Shipping dink mechanics
    case kneeBend
    case contactPosition
    case armStructure
    case headStability
    case followThrough
    // Architected, measured where possible
    case torsoStability
    case balance
    case paddlePosition
    case paddleFaceStability
    case shoulderMovement
    case wristMovement
    case stanceWidth
    case recoveryPosition
    case weightTransfer
    case contactHeight
    case movementConsistency
    case torsoRotation
    case armPath
    case softHands
    case setup

    nonisolated var id: String { rawValue }

    var displayName: String {
        switch self {
        case .kneeBend: return "Knee Bend"
        case .contactPosition: return "Contact Position"
        case .armStructure: return "Arm Structure"
        case .headStability: return "Head Stability"
        case .followThrough: return "Follow-Through"
        case .torsoStability: return "Torso Stability"
        case .balance: return "Balance"
        case .paddlePosition: return "Paddle Position"
        case .paddleFaceStability: return "Paddle-Face Stability"
        case .shoulderMovement: return "Shoulder Movement"
        case .wristMovement: return "Wrist Movement"
        case .stanceWidth: return "Stance Width"
        case .recoveryPosition: return "Recovery Position"
        case .weightTransfer: return "Weight Transfer"
        case .contactHeight: return "Contact Height"
        case .movementConsistency: return "Movement Consistency"
        case .torsoRotation: return "Torso Rotation"
        case .armPath: return "Arm Path"
        case .softHands: return "Soft Hands"
        case .setup: return "Setup"
        }
    }

    /// Short imperative cue the player hears or reads mid-session.
    var shortCue: String {
        switch self {
        case .kneeBend: return "STAY LOW"
        case .contactPosition: return "CONTACT OUT FRONT"
        case .armStructure: return "FIRM ARM SHAPE"
        case .headStability: return "HEAD STILL"
        case .followThrough: return "FINISH THE SWING"
        case .torsoStability: return "QUIET TORSO"
        case .balance: return "STAY BALANCED"
        case .paddlePosition: return "PADDLE UP"
        case .paddleFaceStability: return "STEADY FACE"
        case .shoulderMovement: return "LEAD WITH SHOULDER"
        case .wristMovement: return "QUIET WRIST"
        case .stanceWidth: return "WIDEN BASE"
        case .recoveryPosition: return "RESET YOUR FEET"
        case .weightTransfer: return "MOVE THROUGH IT"
        case .contactHeight: return "LIFT CONTACT"
        case .movementConsistency: return "SAME EVERY REP"
        case .torsoRotation: return "ROTATE THE CHEST"
        case .armPath: return "CLEAN ARM PATH"
        case .softHands: return "SOFT HANDS"
        case .setup: return "SET UP EARLY"
        }
    }
}

/// A single weighted mechanic inside a shot's rubric.
nonisolated struct RubricComponent: Codable, Sendable, Identifiable {
    let mechanic: MechanicID
    /// 0...1, weights within a rubric sum to 1.
    var weight: Double
    /// False while a mechanic is defined but not yet measurable from pose.
    var isMeasured: Bool

    nonisolated var id: String { mechanic.rawValue }
}

/// Per-shot scoring rubric. Every shot has its OWN rubric — never a single
/// generic model across shot types.
nonisolated struct ShotRubric: Codable, Sendable {
    let shot: ShotType
    var components: [RubricComponent]
    /// Bumped whenever weights or benchmarks change, so historical scores
    /// remain interpretable.
    var version: Int

    var measuredComponents: [RubricComponent] { components.filter(\.isMeasured) }

    /// Weights renormalised over the components we can actually measure today.
    var normalizedMeasuredWeights: [MechanicID: Double] {
        let measured = measuredComponents
        let total = measured.reduce(0) { $0 + $1.weight }
        guard total > 0 else { return [:] }
        return Dictionary(uniqueKeysWithValues: measured.map { ($0.mechanic, $0.weight / total) })
    }
}

/// Registry of rubrics. Weights are configurable at runtime (developer mode).
nonisolated enum RubricLibrary {
    nonisolated(unsafe) private static var overrides: [ShotType: ShotRubric] = [:]

    static func rubric(for shot: ShotType) -> ShotRubric {
        if let override = overrides[shot] { return override }
        return defaultRubric(for: shot)
    }

    static func setWeights(_ weights: [MechanicID: Double], for shot: ShotType) {
        var rubric = rubric(for: shot)
        rubric.components = rubric.components.map { component in
            var copy = component
            if let newWeight = weights[component.mechanic] { copy.weight = newWeight }
            return copy
        }
        overrides[shot] = rubric
    }

    static func resetOverrides() { overrides.removeAll() }

    static func defaultRubric(for shot: ShotType) -> ShotRubric {
        switch shot {
        case .forehandDink, .backhandDink:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .kneeBend, weight: 0.25, isMeasured: true),
                RubricComponent(mechanic: .contactPosition, weight: 0.25, isMeasured: true),
                RubricComponent(mechanic: .armStructure, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .headStability, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .followThrough, weight: 0.15, isMeasured: true)
            ], version: 1)

        case .reset:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .balance, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .contactPosition, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .kneeBend, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .softHands, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .headStability, weight: 0.10, isMeasured: true),
                RubricComponent(mechanic: .paddleFaceStability, weight: 0.15, isMeasured: false)
            ], version: 1)

        case .thirdShotDrop:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .kneeBend, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .contactPosition, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .followThrough, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .headStability, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .weightTransfer, weight: 0.25, isMeasured: true)
            ], version: 1)

        case .serve:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .stanceWidth, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .kneeBend, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .weightTransfer, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .contactPosition, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .followThrough, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .balance, weight: 0.15, isMeasured: true)
            ], version: 1)

        case .forehandDrive, .backhandDrive:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .kneeBend, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .torsoRotation, weight: 0.25, isMeasured: false),
                RubricComponent(mechanic: .contactPosition, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .weightTransfer, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .followThrough, weight: 0.20, isMeasured: true)
            ], version: 1)

        case .forehandVolley, .backhandVolley, .block, .rollVolley:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .paddlePosition, weight: 0.25, isMeasured: false),
                RubricComponent(mechanic: .contactPosition, weight: 0.25, isMeasured: true),
                RubricComponent(mechanic: .armStructure, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .balance, weight: 0.15, isMeasured: true),
                RubricComponent(mechanic: .headStability, weight: 0.15, isMeasured: true)
            ], version: 1)

        case .returnOfServe, .speedUp, .overhead, .lob:
            return ShotRubric(shot: shot, components: [
                RubricComponent(mechanic: .setup, weight: 0.20, isMeasured: false),
                RubricComponent(mechanic: .contactPosition, weight: 0.25, isMeasured: true),
                RubricComponent(mechanic: .weightTransfer, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .followThrough, weight: 0.20, isMeasured: true),
                RubricComponent(mechanic: .balance, weight: 0.15, isMeasured: true)
            ], version: 1)
        }
    }
}

/// A measurable range for one mechanic. `ideal` scores 100; the score decays
/// through `acceptable` and bottoms out beyond it.
nonisolated struct BenchmarkRange: Codable, Sendable {
    let idealLow: Double
    let idealHigh: Double
    let acceptableLow: Double
    let acceptableHigh: Double
    let unit: String
    /// Where this range came from — never claim clinical validation.
    let source: BenchmarkSource

    func score(for value: Double) -> Double {
        if value >= idealLow && value <= idealHigh { return 100 }
        if value < idealLow {
            let span = max(0.0001, idealLow - acceptableLow)
            let t = (value - acceptableLow) / span
            return max(10, 45 + 55 * min(1, max(0, t)))
        }
        let span = max(0.0001, acceptableHigh - idealHigh)
        let t = (acceptableHigh - value) / span
        return max(10, 45 + 55 * min(1, max(0, t)))
    }
}

nonisolated enum BenchmarkSource: String, Codable, Sendable {
    /// Derived from coaching heuristics — a starting point, not validated science.
    case coachingHeuristic
    /// Set from reviewed elite footage.
    case eliteFootage
    /// Learned from aggregate Paddle Up usage.
    case aggregateUsage
}

/// Benchmark ranges, stored separately from scoring logic so they can be
/// swapped for coach-validated or data-derived values without touching code
/// that consumes them.
nonisolated enum BenchmarkLibrary {
    static let version = "v1-heuristic-2026.09"

    /// Ranges keyed by shot + mechanic. Values are expressed in the units the
    /// `MechanicsAnalyzer` produces.
    static func range(shot: ShotType, mechanic: MechanicID) -> BenchmarkRange? {
        switch (shot.group, mechanic) {
        // MARK: Dink
        case (.dink, .kneeBend):
            // Interior knee angle in degrees at contact. 180 = straight legs.
            return BenchmarkRange(idealLow: 118, idealHigh: 148, acceptableLow: 95, acceptableHigh: 172,
                                  unit: "°", source: .coachingHeuristic)
        case (.dink, .contactPosition):
            // Horizontal wrist-to-shoulder distance / shoulder width at contact.
            return BenchmarkRange(idealLow: 0.85, idealHigh: 1.55, acceptableLow: 0.25, acceptableHigh: 2.1,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.dink, .armStructure):
            // Elbow angle at contact.
            return BenchmarkRange(idealLow: 122, idealHigh: 158, acceptableLow: 85, acceptableHigh: 178,
                                  unit: "°", source: .coachingHeuristic)
        case (.dink, .headStability):
            // Head travel during the rep / shoulder width. Lower is better.
            return BenchmarkRange(idealLow: 0, idealHigh: 0.22, acceptableLow: -0.1, acceptableHigh: 0.75,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.dink, .followThrough):
            // Wrist path length after contact / shoulder width.
            return BenchmarkRange(idealLow: 0.45, idealHigh: 1.15, acceptableLow: 0.08, acceptableHigh: 1.9,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.dink, .contactHeight):
            return BenchmarkRange(idealLow: -0.15, idealHigh: 0.35, acceptableLow: -0.6, acceptableHigh: 0.85,
                                  unit: "×torso", source: .coachingHeuristic)

        // MARK: Reset
        case (.reset, .kneeBend):
            return BenchmarkRange(idealLow: 110, idealHigh: 142, acceptableLow: 90, acceptableHigh: 168,
                                  unit: "°", source: .coachingHeuristic)
        case (.reset, .contactPosition):
            return BenchmarkRange(idealLow: 0.75, idealHigh: 1.45, acceptableLow: 0.2, acceptableHigh: 2.0,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.reset, .balance):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.18, acceptableLow: -0.1, acceptableHigh: 0.7,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.reset, .softHands):
            // Peak wrist speed through contact, normalised. Lower = softer.
            return BenchmarkRange(idealLow: 0, idealHigh: 1.5, acceptableLow: -0.2, acceptableHigh: 4.2,
                                  unit: "×shoulder/s", source: .coachingHeuristic)
        case (.reset, .headStability):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.24, acceptableLow: -0.1, acceptableHigh: 0.8,
                                  unit: "×shoulder", source: .coachingHeuristic)

        // MARK: Third-shot drop
        case (.thirdShotDrop, .kneeBend):
            return BenchmarkRange(idealLow: 120, idealHigh: 152, acceptableLow: 95, acceptableHigh: 175,
                                  unit: "°", source: .coachingHeuristic)
        case (.thirdShotDrop, .contactPosition):
            return BenchmarkRange(idealLow: 0.9, idealHigh: 1.7, acceptableLow: 0.3, acceptableHigh: 2.3,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.thirdShotDrop, .followThrough):
            return BenchmarkRange(idealLow: 0.6, idealHigh: 1.5, acceptableLow: 0.1, acceptableHigh: 2.3,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.thirdShotDrop, .headStability):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.3, acceptableLow: -0.1, acceptableHigh: 0.9,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.thirdShotDrop, .weightTransfer):
            // Forward hip travel through contact / shoulder width.
            return BenchmarkRange(idealLow: 0.12, idealHigh: 0.6, acceptableLow: -0.15, acceptableHigh: 1.1,
                                  unit: "×shoulder", source: .coachingHeuristic)

        // MARK: Serve
        case (.serve, .stanceWidth):
            return BenchmarkRange(idealLow: 0.95, idealHigh: 1.7, acceptableLow: 0.45, acceptableHigh: 2.4,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.serve, .kneeBend):
            return BenchmarkRange(idealLow: 130, idealHigh: 162, acceptableLow: 100, acceptableHigh: 178,
                                  unit: "°", source: .coachingHeuristic)
        case (.serve, .weightTransfer):
            return BenchmarkRange(idealLow: 0.2, idealHigh: 0.8, acceptableLow: -0.1, acceptableHigh: 1.4,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.serve, .contactPosition):
            return BenchmarkRange(idealLow: 0.7, idealHigh: 1.5, acceptableLow: 0.15, acceptableHigh: 2.2,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.serve, .followThrough):
            return BenchmarkRange(idealLow: 0.8, idealHigh: 1.9, acceptableLow: 0.2, acceptableHigh: 2.8,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (.serve, .balance):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.22, acceptableLow: -0.1, acceptableHigh: 0.8,
                                  unit: "×shoulder", source: .coachingHeuristic)

        // MARK: Generic fallbacks for preview-quality shots
        case (_, .kneeBend):
            return BenchmarkRange(idealLow: 120, idealHigh: 155, acceptableLow: 95, acceptableHigh: 178,
                                  unit: "°", source: .coachingHeuristic)
        case (_, .contactPosition):
            return BenchmarkRange(idealLow: 0.8, idealHigh: 1.6, acceptableLow: 0.2, acceptableHigh: 2.2,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .armStructure):
            return BenchmarkRange(idealLow: 120, idealHigh: 160, acceptableLow: 85, acceptableHigh: 178,
                                  unit: "°", source: .coachingHeuristic)
        case (_, .headStability):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.26, acceptableLow: -0.1, acceptableHigh: 0.85,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .followThrough):
            return BenchmarkRange(idealLow: 0.5, idealHigh: 1.4, acceptableLow: 0.08, acceptableHigh: 2.3,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .balance):
            return BenchmarkRange(idealLow: 0, idealHigh: 0.2, acceptableLow: -0.1, acceptableHigh: 0.8,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .weightTransfer):
            return BenchmarkRange(idealLow: 0.1, idealHigh: 0.7, acceptableLow: -0.2, acceptableHigh: 1.3,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .stanceWidth):
            return BenchmarkRange(idealLow: 0.9, idealHigh: 1.8, acceptableLow: 0.4, acceptableHigh: 2.5,
                                  unit: "×shoulder", source: .coachingHeuristic)
        case (_, .softHands):
            return BenchmarkRange(idealLow: 0, idealHigh: 1.8, acceptableLow: -0.2, acceptableHigh: 4.5,
                                  unit: "×shoulder/s", source: .coachingHeuristic)
        case (_, .contactHeight):
            return BenchmarkRange(idealLow: -0.15, idealHigh: 0.4, acceptableLow: -0.7, acceptableHigh: 1.0,
                                  unit: "×torso", source: .coachingHeuristic)
        default:
            return nil
        }
    }
}

File: web-paddle-up/src/lib/pu/mechanics.ts
/**
 * Mechanic taxonomy, per-shot rubrics (configurable weights) and benchmark
 * ranges. Benchmarks live here, apart from screen logic, so they can later be
 * replaced with values derived from coach review or elite footage.
 */

import { groupOf, type ShotGroup, type ShotType } from "./shots";

export type MechanicID =
  | "kneeBend"
  | "contactPosition"
  | "armStructure"
  | "headStability"
  | "followThrough"
  | "torsoStability"
  | "balance"
  | "paddlePosition"
  | "paddleFaceStability"
  | "shoulderMovement"
  | "wristMovement"
  | "stanceWidth"
  | "recoveryPosition"
  | "weightTransfer"
  | "contactHeight"
  | "movementConsistency"
  | "torsoRotation"
  | "armPath"
  | "softHands"
  | "setup";

export const mechanicName: Record<MechanicID, string> = {
  kneeBend: "Knee Bend",
  contactPosition: "Contact Position",
  armStructure: "Arm Structure",
  headStability: "Head Stability",
  followThrough: "Follow-Through",
  torsoStability: "Torso Stability",
  balance: "Balance",
  paddlePosition: "Paddle Position",
  paddleFaceStability: "Paddle-Face Stability",
  shoulderMovement: "Shoulder Movement",
  wristMovement: "Wrist Movement",
  stanceWidth: "Stance Width",
  recoveryPosition: "Recovery Position",
  weightTransfer: "Weight Transfer",
  contactHeight: "Contact Height",
  movementConsistency: "Movement Consistency",
  torsoRotation: "Torso Rotation",
  armPath: "Arm Path",
  softHands: "Soft Hands",
  setup: "Setup",
};

/** Short imperative cue the player hears or reads mid-session. */
export const mechanicCue: Record<MechanicID, string> = {
  kneeBend: "STAY LOW",
  contactPosition: "CONTACT OUT FRONT",
  armStructure: "FIRM ARM SHAPE",
  headStability: "HEAD STILL",
  followThrough: "FINISH THE SWING",
  torsoStability: "QUIET TORSO",
  balance: "STAY BALANCED",
  paddlePosition: "PADDLE UP",
  paddleFaceStability: "STEADY FACE",
  shoulderMovement: "LEAD WITH SHOULDER",
  wristMovement: "QUIET WRIST",
  stanceWidth: "WIDEN BASE",
  recoveryPosition: "RESET YOUR FEET",
  weightTransfer: "MOVE THROUGH IT",
  contactHeight: "LIFT CONTACT",
  movementConsistency: "SAME EVERY REP",
  torsoRotation: "ROTATE THE CHEST",
  armPath: "CLEAN ARM PATH",
  softHands: "SOFT HANDS",
  setup: "SET UP EARLY",
};

export const allMechanics = Object.keys(mechanicName) as MechanicID[];

export interface RubricComponent {
  mechanic: MechanicID;
  /** 0...1, weights within a rubric sum to 1. */
  weight: number;
  /** False while a mechanic is defined but not yet measurable from pose. */
  isMeasured: boolean;
}

export interface ShotRubric {
  shot: ShotType;
  components: RubricComponent[];
  version: number;
}

function component(
  mechanic: MechanicID,
  weight: number,
  isMeasured: boolean,
): RubricComponent {
  return { mechanic, weight, isMeasured };
}

/**
 * Per-shot scoring rubric. Every shot has its OWN rubric — never a single
 * generic model across shot types.
 */
export function defaultRubric(shot: ShotType): ShotRubric {
  switch (shot) {
    case "forehandDink":
    case "backhandDink":
      return {
        shot,
        version: 1,
        components: [
          component("kneeBend", 0.25, true),
          component("contactPosition", 0.25, true),
          component("armStructure", 0.2, true),
          component("headStability", 0.15, true),
          component("followThrough", 0.15, true),
        ],
      };
    case "reset":
      return {
        shot,
        version: 1,
        components: [
          component("balance", 0.2, true),
          component("contactPosition", 0.2, true),
          component("kneeBend", 0.2, true),
          component("softHands", 0.15, true),
          component("headStability", 0.1, true),
          component("paddleFaceStability", 0.15, false),
        ],
      };
    case "thirdShotDrop":
      return {
        shot,
        version: 1,
        components: [
          component("kneeBend", 0.2, true),
          component("contactPosition", 0.2, true),
          component("followThrough", 0.2, true),
          component("headStability", 0.15, true),
          component("weightTransfer", 0.25, true),
        ],
      };
    case "serve":
      return {
        shot,
        version: 1,
        components: [
          component("stanceWidth", 0.15, true),
          component("kneeBend", 0.15, true),
          component("weightTransfer", 0.2, true),
          component("contactPosition", 0.2, true),
          component("followThrough", 0.15, true),
          component("balance", 0.15, true),
        ],
      };
    case "forehandDrive":
    case "backhandDrive":
      return {
        shot,
        version: 1,
        components: [
          component("kneeBend", 0.15, true),
          component("torsoRotation", 0.25, false),
          component("contactPosition", 0.2, true),
          component("weightTransfer", 0.2, true),
          component("followThrough", 0.2, true),
        ],
      };
    case "forehandVolley":
    case "backhandVolley":
    case "block":
    case "rollVolley":
      return {
        shot,
        version: 1,
        components: [
          component("paddlePosition", 0.25, false),
          component("contactPosition", 0.25, true),
          component("armStructure", 0.2, true),
          component("balance", 0.15, true),
          component("headStability", 0.15, true),
        ],
      };
    case "returnOfServe":
    case "speedUp":
    case "overhead":
    case "lob":
      return {
        shot,
        version: 1,
        components: [
          component("setup", 0.2, false),
          component("contactPosition", 0.25, true),
          component("weightTransfer", 0.2, true),
          component("followThrough", 0.2, true),
          component("balance", 0.15, true),
        ],
      };
  }
}

const overrides = new Map<ShotType, ShotRubric>();

export function rubricFor(shot: ShotType): ShotRubric {
  return overrides.get(shot) ?? defaultRubric(shot);
}

export function setRubricWeights(
  weights: Partial<Record<MechanicID, number>>,
  shot: ShotType,
): void {
  const rubric = rubricFor(shot);
  overrides.set(shot, {
    ...rubric,
    components: rubric.components.map((item) => ({
      ...item,
      weight: weights[item.mechanic] ?? item.weight,
    })),
  });
}

export function resetRubricOverrides(): void {
  overrides.clear();
}

export function measuredComponents(rubric: ShotRubric): RubricComponent[] {
  return rubric.components.filter((item) => item.isMeasured);
}

/** Weights renormalised over the components we can actually measure today. */
export function normalizedMeasuredWeights(
  rubric: ShotRubric,
): Partial<Record<MechanicID, number>> {
  const measured = measuredComponents(rubric);
  const total = measured.reduce((sum, item) => sum + item.weight, 0);
  if (total <= 0) return {};
  const result: Partial<Record<MechanicID, number>> = {};
  for (const item of measured) result[item.mechanic] = item.weight / total;
  return result;
}

/** Where a benchmark range came from — never claim clinical validation. */
export type BenchmarkSource =
  | "coachingHeuristic"
  | "eliteFootage"
  | "aggregateUsage";

export interface BenchmarkRange {
  idealLow: number;
  idealHigh: number;
  acceptableLow: number;
  acceptableHigh: number;
  unit: string;
  source: BenchmarkSource;
}

/** `ideal` scores 100; the score decays through `acceptable` and bottoms out. */
export function scoreForValue(range: BenchmarkRange, value: number): number {
  if (value >= range.idealLow && value <= range.idealHigh) return 100;
  if (value < range.idealLow) {
    const span = Math.max(0.0001, range.idealLow - range.acceptableLow);
    const t = (value - range.acceptableLow) / span;
    return Math.max(10, 45 + 55 * Math.min(1, Math.max(0, t)));
  }
  const span = Math.max(0.0001, range.acceptableHigh - range.idealHigh);
  const t = (range.acceptableHigh - value) / span;
  return Math.max(10, 45 + 55 * Math.min(1, Math.max(0, t)));
}

export const BENCHMARK_VERSION = "v1-heuristic-2026.09";

function range(
  idealLow: number,
  idealHigh: number,
  acceptableLow: number,
  acceptableHigh: number,
  unit: string,
): BenchmarkRange {
  return {
    idealLow,
    idealHigh,
    acceptableLow,
    acceptableHigh,
    unit,
    source: "coachingHeuristic",
  };
}

/**
 * Benchmark ranges, stored separately from scoring logic so they can be
 * swapped for coach-validated or data-derived values without touching code
 * that consumes them.
 */
export function benchmarkRange(
  shot: ShotType,
  mechanic: MechanicID,
): BenchmarkRange | null {
  const group: ShotGroup = groupOf(shot);

  if (group === "dink") {
    switch (mechanic) {
      case "kneeBend":
        return range(118, 148, 95, 172, "°");
      case "contactPosition":
        return range(0.85, 1.55, 0.25, 2.1, "×shoulder");
      case "armStructure":
        return range(122, 158, 85, 178, "°");
      case "headStability":
        return range(0, 0.22, -0.1, 0.75, "×shoulder");
      case "followThrough":
        return range(0.45, 1.15, 0.08, 1.9, "×shoulder");
      case "contactHeight":
        return range(-0.15, 0.35, -0.6, 0.85, "×torso");
    }
  }

  if (group === "reset") {
    switch (mechanic) {
      case "kneeBend":
        return range(110, 142, 90, 168, "°");
      case "contactPosition":
        return range(0.75, 1.45, 0.2, 2.0, "×shoulder");
      case "balance":
        return range(0, 0.18, -0.1, 0.7, "×shoulder");
      case "softHands":
        return range(0, 1.5, -0.2, 4.2, "×shoulder/s");
      case "headStability":
        return range(0, 0.24, -0.1, 0.8, "×shoulder");
    }
  }

  if (group === "thirdShotDrop") {
    switch (mechanic) {
      case "kneeBend":
        return range(120, 152, 95, 175, "°");
      case "contactPosition":
        return range(0.9, 1.7, 0.3, 2.3, "×shoulder");
      case "followThrough":
        return range(0.6, 1.5, 0.1, 2.3, "×shoulder");
      case "headStability":
        return range(0, 0.3, -0.1, 0.9, "×shoulder");
      case "weightTransfer":
        return range(0.12, 0.6, -0.15, 1.1, "×shoulder");
    }
  }

  if (group === "serve") {
    switch (mechanic) {
      case "stanceWidth":
        return range(0.95, 1.7, 0.45, 2.4, "×shoulder");
      case "kneeBend":
        return range(130, 162, 100, 178, "°");
      case "weightTransfer":
        return range(0.2, 0.8, -0.1, 1.4, "×shoulder");
      case "contactPosition":
        return range(0.7, 1.5, 0.15, 2.2, "×shoulder");
      case "followThrough":
        return range(0.8, 1.9, 0.2, 2.8, "×shoulder");
      case "balance":
        return range(0, 0.22, -0.1, 0.8, "×shoulder");
    }
  }

  // Generic fallbacks for preview-quality shots.
  switch (mechanic) {
    case "kneeBend":
      return range(120, 155, 95, 178, "°");
    case "contactPosition":
      return range(0.8, 1.6, 0.2, 2.2, "×shoulder");
    case "armStructure":
      return range(120, 160, 85, 178, "°");
    case "headStability":
      return range(0, 0.26, -0.1, 0.85, "×shoulder");
    case "followThrough":
      return range(0.5, 1.4, 0.08, 2.3, "×shoulder");
    case "balance":
      return range(0, 0.2, -0.1, 0.8, "×shoulder");
    case "weightTransfer":
      return range(0.1, 0.7, -0.2, 1.3, "×shoulder");
    case "stanceWidth":
      return range(0.9, 1.8, 0.4, 2.5, "×shoulder");
    case "softHands":
      return range(0, 1.8, -0.2, 4.5, "×shoulder/s");
    case "contactHeight":
      return range(-0.15, 0.4, -0.7, 1.0, "×torso");
    default:
      return null;
  }
}
